app/src/org/commcare/android/nfc/NfcActivity.java

```diff
@@ -0,0 +1,150 @@
+import android.content.Intent;
+import android.content.IntentFilter;
+import android.nfc.NfcAdapter;
+import android.nfc.Tag;
+import android.os.Build;
+import android.os.Bundle;
+import android.widget.TextView;
+import android.widget.Toast;
+
+import org.commcare.android.logging.ReportingUtils;
+import org.commcare.dalvik.R;
+import org.javarosa.core.services.locale.Localization;
+
+/**
+ * Parent activity that provides all of the functionality common to any NFC action that CommCare
+ * supports
+ *
+ * @author Aliza Stone
+ */
+public abstract class NfcActivity extends Activity {
+
+    protected static final String NFC_PAYLOAD_MULT_TYPES_ARG = "types";
+    protected static final String NFC_PAYLOAD_SINGLE_TYPE_ARG = "type";
+    protected static final String NFC_DOMAIN_ARG = "domain";
+
+    protected NfcManager nfcManager;
+    protected PendingIntent pendingNfcIntent;
+    protected String domainForType;
+
+    @Override
+    protected void onCreate(Bundle savedInstanceState) {
+        super.onCreate(savedInstanceState);
+        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.JELLY_BEAN) {
+            initFields();
+            createPendingRestartIntent();
+            setContentView(R.layout.nfc_instructions_view);
+            ((TextView)findViewById(R.id.nfc_instructions_text_view)).
+                    setText(Localization.get(getInstructionsTextKey()));
+        }
+    }
+
+    protected void initFields() {
+        this.nfcManager = new NfcManager(this);
+
+        this.domainForType = getIntent().getStringExtra(NFC_DOMAIN_ARG);
+        if (this.domainForType == null) {
+            // Usually there is no reason for the user to provide this themselves, and we will just
+            // set it to be the project's domain name
+            this.domainForType = ReportingUtils.getDomain();
+        }
+    }
+    /**
+     * Create an intent for restarting this activity, which will be passed to enableForegroundDispatch(),
+     * thus instructing Android to start the intent when the device detects a new NFC tag. Adding
+     * FLAG_ACTIVITY_SINGLE_TOP makes it so that onNewIntent() can be called in this activity when
+     * the intent is started.
+     **/
+    protected void createPendingRestartIntent() {
+        Intent i = new Intent(this, getClass());
+        i.addFlags(Intent.FLAG_ACTIVITY_SINGLE_TOP);
+        this.pendingNfcIntent = PendingIntent.getActivity(this, 0, i, 0);
+    }
+
+    @Override
+    protected void onResume() {
+        super.onResume();
+
+        if (Build.VERSION.SDK_INT < Build.VERSION_CODES.JELLY_BEAN) {
+            finishWithErrorToast("nfc.min.version.message");
+            return;
+        }
+
+        try {
+            nfcManager.checkForNFCSupport();
+            if (requiredFieldsMissing()) {
+                return;
+            }
+            setReadyToHandleTag();
+        } catch (NfcManager.NfcNotEnabledException e) {
+            finishWithErrorToast("nfc.not.enabled");
```

Comment:
In this case we could prompt them to turn on NFC and listen for them to activate it, no?

Reply:
Mm good call that I should add some instructions to activate it to the error message, but I think listening for activation goes beyond what is reasonably needed, and also wouldn't really fit into the control flow. Throwing this exception will finish the NFC activity, and then they can simply re-press the button that triggers the activity to open once they've gone into settings and activated it.